### PR TITLE
[SITES-590] Feature/2fa

### PR DESCRIPTION
**Pull request by ltankey**
**COMMENTS WELCOME**

This is bigger than I intended, because it essentially became an epic.

The primary goal of this PR was to introduce 2FA to the application via a Devise extension called `two_factor_authentication`. However, as this gem only handles authentication (and not subsequent identity verification or 2FA setup), there are additional functions included to support this (that aren't particularly DRY just yet).

The functions include:

* sign-in with 2FA (TOTP or OTP)
* initialise 2FA for accounts to complete setup
* verify identity via 2FA enforced with before_action calls 

The components include:

- [x] SMSService (as the name suggests)
- [x] one-time-password service for generating one-time-passwords (and corresponding _sent_at) values, along with validation methods
- [x] two_factor_authentication extension for Devise
- [x] 2FA-verification protection of sensitive fields

Still to do:
- [ ] google authenticator setup (qr code generation)
- [ ] limit one-time-password attempts/retries

### CODE REVIEW

File: app/controllers/users/two_factor_verification_controller.rb [hunk 1]
```diff
@@ -0,0 +1,57 @@
+module Users
+  class TwoFactorVerificationController < ApplicationController
+    before_action :check_direct_otp_expiry, only: [:show, :create]
+
+
+    def show
```
Comment (freshtonic) [inline app/controllers/users/two_factor_verification_controller.rb]: Should be a POST (create).

File: app/controllers/users/two_factor_verification_controller.rb [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+      unless current_user.totp_enabled?
+        current_user.create_direct_otp
+        r = SMSService.new.send_two_factor_confirmation_code!(current_user)
+
+        # TODO: make this async and handle service failure
+        unless r.code == 202
+          raise 'There was an error sending the SMS'
```
Comment (freshtonic) [inline app/controllers/users/two_factor_verification_controller.rb]: Should probably return a 400 ?

File: app/controllers/users/two_factor_verification_controller.rb [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        end
+      end
+    end
+
+    def create
```
Comment (freshtonic) [inline app/controllers/users/two_factor_verification_controller.rb]: Oh but there's a create here. Maybe there should be a different endpoint for sending the SMS, above?

Comment (ltankey) [inline app/controllers/users/two_factor_verification_controller.rb]: This comes down to how we expect the SMS service to work. It's currently modelled off the `two_factor_authentication` gem which essentially sends it on `#show`.

This has an advantage in that it doesn't require the user have to click through to have an SMS sent to them - however the disadvantage is it doesn't provide the user with a warning that 2FA will be required, or give them an opportunity to back out of it.

We could choose to use the show action to display an informative screen outlining that verification is required via 2FA, and get the user to click through (post to create) to send the SMS. The subsequent call to submit the code could be an update action.

File: app/controllers/users/two_factor_verification_controller.rb (hunk 2 above)
Comment (ltankey) [inline app/controllers/users/two_factor_verification_controller.rb]: We probably do - but do we need to handle this in any sophisticated manner? If the SMS API is (intermittently or otherwise) down, do we want a strategy of retrying? If not, we could probably settle for displaying an informative message of `The 2FA service is currently down`, but would need to test how helpful that is.

Comment (freshtonic) [inline app/controllers/users/two_factor_verification_controller.rb]: As we discussed on slack, fire the SMS request off in a new thread and ignore it if it fails to send, but indicate to the user that they can try again if they do not receive an SMS.

File: app/controllers/users/two_factor_verification_controller.rb (hunk 3 above)
Comment (freshtonic) [inline app/controllers/users/two_factor_verification_controller.rb]: The problem with `show` is that it's `GET` with a side-effect. How are you making sure that link does not get crawled?

Comment (ltankey) [inline app/controllers/users/two_factor_verification_controller.rb]: Really good point. I guess it'd be stopped at the outer by needing to be signed in, but your point still stands.

Comment (freshtonic) [inline app/controllers/users/two_factor_verification_controller.rb]: If you do end up deciding to keep the GET, you can add `rel="nofollow"` on links.

Comment (freshtonic) [inline app/controllers/users/two_factor_verification_controller.rb]: And also add that route to `robots.txt`

File: app/controllers/application_controller.rb [hunk 4]
```diff
@@ -15,17 +16,41 @@ class ApplicationController < ActionController::Base
...
+    #   User.includes(:roles).find(user.id)
+    # end
   end
 
   def decorated_current_user
     current_user.try(:decorate)
   end
 
+  #TODO: move this into a Warden after_authentication method or
+  # something likely to be executed _after_ TFA has been skipped
```
Comment (tessereth) [inline app/controllers/application_controller.rb]: This comment is confusing. Can you be clearer about what you mean?

File: app/controllers/application_controller.rb [hunk 5]
```diff
@@ -15,17 +16,41 @@ class ApplicationController < ActionController::Base
...
+  def complete_two_factor_setup
+    if current_user && !current_user.account_verified && current_user.confirmed_at
+      # Users need to be able to verify their tfa code
+      #unless request.path.start_with?('/users')
+      if request.path.match(/\A\/users(?!\/edit).*\z/) == nil
```
Comment (tessereth) [inline app/controllers/application_controller.rb]: This regex is confusing. If there's only a few paths that _don't_ require a redirect, it may make more sense to just white list them rather than making assumptions about routing.

File: app/controllers/users/two_factor_setup_controller.rb [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+    def create
+      current_user.unconfirmed_phone_number =  params[:phone_number]
+
+      if current_user.save
+        current_user.send_new_direct_otp_for(:unconfirmed_phone_number)
+        redirect_to confirm_users_two_factor_setup_path and return
+      end
+
+      render :new
```
Comment (tessereth) [inline app/controllers/users/two_factor_setup_controller.rb]: Stylistically, we've tended to put the `render :new` in an else clause rather than having an early return.

File: app/services/otp_service.rb [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+  end
+
+
+  def create_otp(field)
+    create_direct_otp(field)
+  end
+
+
+  def authenticate_otp(code_field, code)
+    direct_otp = @user.send(code_field)
```
Comment (tessereth) [inline app/services/otp_service.rb]: May be worth adding a comment to this method "Never under any circumstances call with a user provided code_field argument" or something to that effect.

File: app/validators/aus_phone_number_validator.rb [hunk 8]
```diff
@@ -0,0 +1,7 @@
+class AusPhoneNumberValidator < ActiveModel::EachValidator
+  def validate_each(record, attribute, value)
+    if value != nil && value !~ /\A04[0-9]{8}\z/
```
Comment (tessereth) [inline app/validators/aus_phone_number_validator.rb]: This is a little pedantic. It would be much nicer to allow the user to put spaces in their number and just strip them server side.

Comment (ltankey) [inline app/validators/aus_phone_number_validator.rb]: Agreed, although I think there's a whole flow around setting up phone numbers and authenticator details that's likely going to be left for another PR. I think we will address this validator concern as part of that.

File: app/views/devise/two_factor_authentication/show.html.haml [hunk 9]
```diff
@@ -0,0 +1,28 @@
+%artcile.sign-in
```
Comment (tessereth) [inline app/views/devise/two_factor_authentication/show.html.haml]: Misspelling of article?

**Comment by tessereth** [comment]
Missing tests but otherwise LGTM

**Comment by tessereth** [comment]
Update changelog post 1.1.0 release

File: CHANGELOG.md [hunk 10]
```diff
@@ -8,6 +8,7 @@ This file is influenced by http://keepachangelog.com/.
 - Added collecting test metadata for CircleCI
 - Added ability to mark pages as placeholders that appear in navigation
 - Rake task for dumping, santising, and uploading the database to AWS S3
+- Added two-factor authentication to sign-in/up and verification
```
Comment (maxious) [inline CHANGELOG.md]: will need v1.1.0 rebase

File: app/views/users/two_factor_setup/confirm.html.haml [hunk 11]
```diff
@@ -0,0 +1,23 @@
+%artcile.verify
```
Comment (maxious) [inline app/views/users/two_factor_setup/confirm.html.haml]: Misspelling article might have bad effects on UI Kit styling.

File: bin/scrub.rb [hunk 12]
```diff
@@ -33,6 +33,16 @@
       parts[col_index[:last_name]] = 'Nameless'
       parts[col_index[:confirmation_token]] = '\N'
       parts[col_index[:unconfirmed_email]] = '\N'
+      parts[col_index[:phone_number]] = '0412345678'
```
Comment (maxious) [inline bin/scrub.rb]: perhaps just REDACTED would be easier to read? Took me a sec to understand.

**Comment by maxious** [comment]
Minor tweaks req. but LGTM
